components/tidb_query/src/batch/runner.rs

```diff
@@ -440,16 +424,126 @@ impl<SS: 'static> BatchExecutorsRunner<SS> {
             }
 
             // Grow batch size
-            if batch_size < BATCH_MAX_SIZE {
-                batch_size *= BATCH_GROW_FACTOR;
-                if batch_size > BATCH_MAX_SIZE {
-                    batch_size = BATCH_MAX_SIZE
-                }
-            }
+            grow_batch_size(&mut batch_size);
         }
     }
 
     pub fn collect_storage_stats(&mut self, dest: &mut SS) {
         self.out_most_executor.collect_storage_stats(dest);
     }
+
+    pub fn handle_streaming_request(
+        &mut self,
+    ) -> Result<(Option<(StreamResponse, IntervalRange)>, bool)> {
+        let mut warnings = self.config.new_eval_warnings();
+
+        // if the data read finished, is_drained means "finished"
+        let (mut record_len, mut is_drained) = (0, false);
+        let mut chunk = Chunk::default();
+
+        // record count less than batch size and is not drained
+        while record_len < self.stream_min_rows_each_iter && !is_drained {
+            let (drained, cnt) =
+                self.internal_handle_request(self.stream_batch_size, &mut chunk, &mut warnings)?;
+            record_len += cnt;
+            is_drained = drained;
+
+            // Grow batch size
+            grow_batch_size(&mut self.stream_batch_size)
+        }
+
+        if !is_drained || record_len > 0 {
+            let range = self.out_most_executor.take_scanned_range();
+            return self
+                .make_stream_response(chunk, warnings)
+                .map(|r| (Some((r, range)), is_drained));
+        }
+        Ok((None, true))
+    }
+
+    fn make_stream_response(
+        &mut self,
+        chunk: Chunk,
+        warnings: EvalWarnings,
+    ) -> Result<StreamResponse> {
+        self.out_most_executor
+            .collect_exec_stats(&mut self.exec_stats);
+
+        let mut s_resp = StreamResponse::default();
+        s_resp.set_data(box_try!(chunk.write_to_bytes()));
+
+        s_resp.set_output_counts(
+            self.exec_stats
+                .scanned_rows_per_range
+                .iter()
+                .map(|v| *v as i64)
+                .collect(),
+        );
+
+        s_resp.set_warnings(warnings.warnings.into());
+        s_resp.set_warning_count(warnings.warning_cnt as i64);
+
+        self.exec_stats.clear();
+
+        Ok(s_resp)
+    }
+
+    /// return (is_drained, EvalWarnings, record_cnt)
+    fn internal_handle_request(
+        &mut self,
+        batch_size: usize,
+        chunk: &mut Chunk,
+        warnings: &mut EvalWarnings,
+    ) -> Result<(bool, usize)> {
+        let is_drained;
+        let mut record_cnt = 0;
+
+        self.deadline.check()?;
+
+        let mut result = self.out_most_executor.next_batch(batch_size);
+
+        // fill is_drained
```

Comment:
Incorrect comment: but you are not filling `is_drained` when `result.is_drained` is `Err(_)`